### PR TITLE
Refactoring content cache functionality

### PR DESCRIPTION
**Pull request by devanych**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #5, #10, #22, #91

Removed `DynamicContentAwareTrait` and `DynamicContentAwareInterface`. Removed all functionality from view classes. The view classes now do not know anything about content caching at all.

You can cache a piece of code out of the box like this:

```php
use Yiisoft\View\Cache\CachedContent;
use Yiisoft\View\Cache\DynamicContent;

/**
 * @var \Yiisoft\View\BaseView $view
 * @var \Yiisoft\Cache\CacheInterface $cache
 */

$dynamicContent = new DynamicContent('dynamic-id', static function (array $parameters): string {
    return strtoupper("{$parameters['a']} - {$parameters['b']}");
}, ['a' => 'string-a', 'b' => 'string-b']);

$cachedContent = new CachedContent('cache-id', $cache, [$dynamicContent]);

// In the content, we set: `$dynamicContent->placeholder()`
$fragment = $view->renderFile('path/to/file', ['dynamicContent' => $dynamicContent]);

$fragmentCache = $cachedContent->cache($fragment, 60, new TagDependency('tag-name')); 

$content = $view->render('view/name', ['fragmentCache' => $fragmentCache]);
```

Cache the entire page in the same way:

```php
use Yiisoft\View\Cache\CachedContent;
use Yiisoft\View\Cache\DynamicContent;

/**
 * @var \Yiisoft\View\BaseView $view
 * @var \Yiisoft\Cache\CacheInterface $cache
 */

$dynamicContent = new DynamicContent('dynamic-id', static function (array $parameters): string {
    return strtoupper("{$parameters['a']} - {$parameters['b']}");
}, ['a' => 'string-a', 'b' => 'string-b']);

$cachedContent = new CacheContent('cache-id', $cache, [$dynamicContent]);

$content = $view->render('view/name', ['dynamicContent' => $dynamicContent]);

return $cachedContent->cache($content, 60, new TagDependency('tag-name'));
```

When using the `yiisoft/yii-widgets` package, it will be like this:

```php
use Yiisoft\View\Cache\DynamicContent;
use Yiisoft\Yii\Widgets\FragmentCache;

$dynamicContent = new DynamicContent('dynamic-id', static function (array $parameters): string {
    return strtoupper("{$parameters['a']} - {$parameters['b']}");
}, ['a' => 'string-a', 'b' => 'string-b']);

$fragmentCache = FragmentCache::widget()->id('cache-id')->dynamicContents($dynamicContent);

// Output:
$fragmentCache->begin();
    echo 'Content to be cached ...';
    echo $dynamicContent->placeholder();
    echo 'Content to be cached ...';
$fragmentCache::end();
```

### CODE REVIEW

**Comment by terabytesoftw** [comment]
@devanych What do you think of moving to a new cache package, basically they are two independent caches of the view, which are used by the FramentCache::class widget.

**Review by terabytesoftw — APPROVED** [review]

**Comment by devanych** [comment]
> @devanych What do you think of moving to a new cache package, basically they are two independent caches of the view, which are used by the `FramentCache::class` widget.

Do you mean to move `CacheContent` and `DynamicContent` to the `yiisoft/cache` package?

**Comment by terabytesoftw** [comment]
> > @devanych ¿Qué piensas de pasar a un nuevo paquete de caché, básicamente son dos cachés independientes de la vista, que son utilizados por el widget.`FramentCache::class`
> 
> ¿Te refieres a moverte y al paquete?`CacheContent``DynamicContent``yiisoft/cache`

yes

**Comment by devanych** [comment]
> yes, `yiisoft/cache-view`, i think in the future we could add more functionalities in view.

In principle, can transfer it. @samdark what do you think?

**Comment by samdark** [comment]
I'm not sure it's usable without view.

**Comment by devanych** [comment]
> I'm not sure it's usable without view.

Yes, it probably is.

Moved the "CacheContent" and "DynamicContent" classes to the "Cache" subfolder. If we need additional classes for caching content, add them to this subfolder.

File: src/Cache/CacheContent.php [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+ * CacheContent caches content, supports the use of dynamic content {@see DynamicContent} inside cached content.
+ */
+final class CacheContent
+{
+    private string $id;
+    private CacheInterface $cache;
+    private CacheKeyNormalizer $cacheKeyNormalizer;
+
+    /**
+     * @var array<string, DynamicContent>
```
Comment (samdark) [inline src/Cache/CacheContent.php:31]:
```suggestion
     * @var DynamicContent[]
```

File: src/Cache/CacheContent.php [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    /**
+     * @var string[]
+     */
+    private array $variations = [];
+
+    /**
+     * @param string $id The unique identifier of the cached content.
+     * @param CacheInterface $cache The cache instance.
+     * @param array $dynamicContents The dynamic content instances.
```
Comment (samdark) [inline src/Cache/CacheContent.php:43]:
```suggestion
     * @param DynamicContent[] $dynamicContents The dynamic content instances.
```

File: src/Cache/CacheContent.php [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            $this->cache->getOrSet($this->cacheKey(), static fn (): string => $content, $ttl, $dependency, $beta),
+        );
+    }
+
+    /**
+     * Returns cached content with placeholders replaced with actual dynamic content.
+     *
+     * @return string|null The cached content. Null is returned if valid content is not found in the cache.
+     */
+    public function cachedContent(): ?string
```
Comment (samdark) [inline src/Cache/CacheContent.php:79]: The name here doesn't reflect what's actually returned. Suggestions:

- getFinalContent
- getContentString

File: src/Cache/CacheContent.php [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+     * @param string $content The content of the item to cache store.
+     * @param DateInterval|int|null $ttl The TTL of the cached content.
+     * @param Dependency|null $dependency The dependency of the cached content.
+     * @param float $beta The value for calculating the range that is used for "Probably early expiration".
+     *
+     * @see CacheInterface::getOrSet()
+     *
+     * @return string The rendered cached content.
+     */
+    public function cache(string $content, $ttl = 60, Dependency $dependency = null, float $beta = 1.0): string
```
Comment (samdark) [inline src/Cache/CacheContent.php:67]: Does it make sense to call it multiple times? If not, maybe $content should be part of the constructor and the rest should be additional configuration methods?

File: src/Cache/CacheContent.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+use function gettype;
+use function is_string;
+use function is_object;
+use function sprintf;
+use function strtr;
+
+/**
+ * CacheContent caches content, supports the use of dynamic content {@see DynamicContent} inside cached content.
+ */
+final class CacheContent
```
Comment (samdark) [inline src/Cache/CacheContent.php:24]: Cache content sounds like an action, maybe:

```suggestion
final class ContentCache
```

File: src/Cache/CacheContent.php (hunk 1 above)
Comment (devanych) [reply src/Cache/CacheContent.php:31]: You can add one object several times and replace placeholders with content several times.
https://github.com/yiisoft/view/pull/159/files#diff-ab378c33bf6ef55354f2cbd7d49a80ec83ef0ff0e3c96b7ff364c58e5b068380R116

File: src/Cache/CacheContent.php (hunk 3 above)
Comment (devanych) [reply src/Cache/CacheContent.php:79]: It returns cached content, if there is no content in the cache, it returns `null`. It seems to me that `cachedContent` is much clearer in this case.

File: src/Cache/CacheContent.php (hunk 4 above)
Comment (devanych) [reply src/Cache/CacheContent.php:67]: The `CacheContent ' object can be used inside a view file. And we need to have the placeholders separately, for use in the view file.

File: src/Cache/CacheContent.php (hunk 5 above)
Comment (devanych) [reply src/Cache/CacheContent.php:24]: Then instead of `DynamicContent` need `ContentDynamic`?

Comment (samdark) [reply src/Cache/CacheContent.php:24]: No.

File: src/Cache/CacheContent.php (hunk 3 above)
Comment (samdark) [reply src/Cache/CacheContent.php:79]: It returns cached content with placeholders replaced. That's not what we set as content in `cache()`.

**Pull request merged**